### ISSUE TITLE
Roslyn fails to detect null assignments inside 'foreach'

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#69989]
**Version Used**: 

```
Microsoft Visual Studio Professional 2022
Version 17.8.0 Preview 1.0
VisualStudio.17.Preview/17.8.0-pre.1.0+34004.107
Microsoft .NET Framework
Version 4.8.09037
```

**Steps to Reproduce**:

Compile and run the following code:
```cs
using System;
using System.Collections.Generic;

#nullable enable
internal class Program
{
  static void Main(string[] args)
  {
    Foreach();
  }

  static void Foreach()
  {
    string[]? strings = new string[] { "", "" };
    object[] objects = new object[] { new(), new() };
    HashSet<string> hashSet = new HashSet<string>();

    foreach (object _ in objects)
    {
      strings = null;
    }

    hashSet.UnionWith(strings);
  }
}
```

**Expected Behavior**:

`CS604: Possible null reference argument` is reported for `hashSet.UnionWith(strings);`

**Actual Behavior**:

No warnings. The code fails at runtime with an `ArgumentNullException`
Roslyn also claims that `strings` is not null at the last line of `Foreach` method when hovered.

**Notes**:
Interestingly, if you convert the cycle to `for` Roslyn will detect the nullable state:
```
  static void For()
  {
    string[]? strings = new string[] { "", "" };
    object[] objects = new object[] { new(), new() };
    HashSet<string> hashSet = new HashSet<string>();

    for (var index = 0; index < objects.Length; index++)
    {
      var _ = objects[index];
      strings = null;
    }

    hashSet.UnionWith(strings); // correct warning
  }
  ```
